src/types/timestamp.rs

```diff
@@ -134,8 +125,16 @@ impl Timestamp {
             // This timestamp stores its fractional seconds in its `date_time` field.
             // We'll need to convert the date_time's nanoseconds to a Decimal and return it.
             Some(Digits(number_of_digits)) => {
-                let coefficient = first_n_digits_of(*number_of_digits, self.date_time.nanosecond());
-                let exponent = -((coefficient as f64).log10().ceil() as i64);
+                const MAX_NANOSECOND_DIGITS: u32 = 9; // If it were 10, it'd be > a second
+                let nanoseconds = self.date_time.nanosecond();
+                let leading_zeros = MAX_NANOSECOND_DIGITS
+                    - super::num_decimal_digits_in_u64(nanoseconds as u64) as u32;
+                let coefficient = if leading_zeros >= *number_of_digits {
+                    0
+                } else {
+                    first_n_digits_of(*number_of_digits - leading_zeros, nanoseconds)
+                };
+                let exponent = -i64::from(*number_of_digits);
```

Comment:
🗺️ When comparing two `Timestamp` values: if one timestamp has a `Decimal` fractional seconds and the other stores its fractional seconds as nanoseconds on the `DateTime`, the `eq` method will upconvert the nanoseconds fractional seconds to a `Decimal` for the sake of comparison.

The original implementation was overly simple; `ion-tests` shed light on a few corner cases that weren't handled properly.